### PR TITLE
DS-3741 hibernate deprecation fix applied feedback

### PR DESCRIPTION
**Pull request by Raf-atmire**
Made this new PR in order to have a new overview of possible feedback and fixes to the codebase without destroying the timeline and comments on the previous PR that can be found here: https://github.com/DSpace/DSpace/pull/1964
Note that new development and adaptions will be done on this PR instead of the old one to preserve some kind of history and overview.

DS-3741: https://github.com/DSpace/DSpace/issues/7088

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/checker/dao/impl/MostRecentChecksumDAOImpl.java
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/checker/dao/impl/MostRecentChecksumDAOImpl.java]: Does this work?  I haven't found anything to say that `MostRecentChecksum_.bitstream` means `Bitstream.id`.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/CommunityDAOImpl.java
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/content/dao/impl/CommunityDAOImpl.java]: Couldn't this just be 'Predicate[] orPredicate = new Predicate[actions.length]'?  Then we don't have to create more work and more garbage by calculating toArray() later.  IIRC there's similar code in CollectionDAOImpl.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java]: I don't see where offset and limit are applied, in the new code.

File: dspace-api/src/main/java/org/dspace/core/AbstractHibernateDAO.java
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/core/AbstractHibernateDAO.java]: Do we need this?

File: dspace-api/src/main/java/org/dspace/eperson/dao/impl/SubscriptionDAOImpl.java
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/eperson/dao/impl/SubscriptionDAOImpl.java]: Could not all of these 1- and 2-element Order lists be replaced by using the varargs form of CriteriaQuery.orderBy()?

File: dspace-api/src/main/java/org/dspace/harvest/dao/impl/HarvestedCollectionDAOImpl.java
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/harvest/dao/impl/HarvestedCollectionDAOImpl.java]: We seem to have lost 'limit'.

**Comment by mwoodiupui** [comment]
Overall I like this change.  And I've run all unit and integration tests on it without complaints.

File: dspace-api/src/main/java/org/dspace/checker/dao/impl/MostRecentChecksumDAOImpl.java
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/checker/dao/impl/MostRecentChecksumDAOImpl.java]: The MostRecentChecksum_ class will make static fields that are named the same as they are in the model class. So the `MostRecentChecksum_.bitstream` in this instance refers to the `private Bitstream bitstream` field in MostRecentChecksum.java which in turn refers to the bitstream_id column. If my interpretation/deduction is correct, this should mean that the bitstream id is used.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/CommunityDAOImpl.java
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/CommunityDAOImpl.java]: That's a good point. That should be possible at first sight and there are indeed multiple cases where this could be implemented like you proposed.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java]: The offset and limit are applied when calling the list method in the return line of this findByMetadataQuery method: https://github.com/DSpace/DSpace/pull/2111/files/316d9f08e577fe8cbee7b3ca850c1099959da6c8#diff-47280133d16cde27773db271f5386298R244
However thank you for asking about this because I noticed that I accidentally swapped the limit and offset around initially, this should be fixed in the newest commit.

File: dspace-api/src/main/java/org/dspace/core/AbstractHibernateDAO.java
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/core/AbstractHibernateDAO.java]: There is indeed no reason for that import to be sat there in comments. I will remove this line entirely in the next commit.

File: dspace-api/src/main/java/org/dspace/harvest/dao/impl/HarvestedCollectionDAOImpl.java
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/harvest/dao/impl/HarvestedCollectionDAOImpl.java]: The way that this method is currently structured by returning a criteriaQuery implies that this method should not set a limit on the results, that has to be done later down the line when a `Query` is made from the `CriteriaQuery` object that this method will return. 
However this method is never used in the current DSpace code. I'll remove the limit parameter and leave the method as is unless we want to return a list of HarvestCollections here instead of the CriteriaQuery, in which case we can call the logic to execute the query and pass along the limit.

File: dspace-api/src/main/java/org/dspace/eperson/dao/impl/SubscriptionDAOImpl.java
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/eperson/dao/impl/SubscriptionDAOImpl.java]: That's a good point, I'm fairly certain that that is possible. Personally I find this easier to read but if using the varargs is simpler/easier then I will refactor the instances of orderBy to use the vargs instead.

**Review by mwoodiupui — APPROVED** [review]
+1  My comments are all addressed, and I think this is overall a worthwhile improvement.  Tests pass.

**Review by tomdesair — APPROVED** [review]
I think it's important that we merge this as soon as possible in order to discovery any query-conversion issues as soon as possible in the current 7.0 development process.

File: dspace/config/modules/rest.cfg
Comment (tdonohue) [inline dspace/config/modules/rest.cfg]: This seems to be an accidental change.  Why is this configuration being removed in this PR?

**Review by tdonohue — CHANGES_REQUESTED** [review]
Overall, this PR looks very good.  However, I've found several inconsistencies (likely small bugs) between the code here and the original code.  See below comments.  

Once these are fixed, I think this is good to merge.  (Update: this also does have a minor merge conflict to resolve)

**Comment by tomdesair** [comment]
@tdonohue Thanks for the thorough review.

I think the `regex.rest-clause` was removed because JPA does not support this anymore. But I'll double check with Raf.

Your other remarks seem valid as well so we'll do another iteration.

**Comment by tdonohue** [comment]
@tomdesair : just a note on the `rest.regex-clause` configuration.  We shouldn't remove it until the code that relies on it is also removed.  I see that the code that uses this configuration still exists on your branch (and in this PR): https://github.com/atmire/DSpace/blob/6603826a8144de2eebcfb7fc7369a8c0bf100282/dspace-rest/src/main/java/org/dspace/rest/FilteredItemsResource.java#L136

Assuming this configuration can no longer be supported & in order to hurry this PR along, I'd recommend undoing the removal of `rest.regex-clause` config, and logging a bug about it in JIRA.  Then, you could create a followup PR to remove it and its related code.  It seems like a separate issue anyways, and it should be logged in JIRA as being removed (so that sites that rely on it realized it is removed).

Beyond that, as soon as the other fixes are applied to this PR, I think the rest of the code looks great.  As I will be out of the office much of next week, I'd encourage a Committer to override my "request for changes" as soon as all my inline comments are resolved.  Otherwise, I'll get back to this the following week.

File: dspace/config/modules/rest.cfg
Comment (Raf-atmire) [reply dspace/config/modules/rest.cfg]: JPA2 does not support SQL restrictions: https://stackoverflow.com/questions/5182701/equivalent-of-hibernates-restrictions-sqlrestriction-in-jpa2-criteria-api

DSpace 6 is using one instance of this property to support a regex check in a query: https://github.com/DSpace/DSpace/blob/dspace-6.3/dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java#L188

@terrywbrady is it a problem that we remove regex restriction support in the old rest api?

Comment (terrywbrady) [reply dspace/config/modules/rest.cfg]: @Raf-atmire , this feature was added to allow non-postgres instances to use the rest report tools.

Are you also removing the code where this property is used?  If so, the rest report tools will break as a part of these changes.

**Review by terrywbrady — CHANGES_REQUESTED** [review]
Once the merge issues are resolved, please ping me and I will re-test the DSpace 6 Rest Reports.  I hope we can continue to preserver the functionality implemented in those tools.  I will do a code review at that time as well.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java
Comment (terrywbrady) [inline dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java]: The DSpace Rest Report tools rely extensively on the matches and doesnt_match operators.  Without those operators, most of the benefits of those tools are lost.  See https://terrywbrady.github.io/restReportTutorial/intro

**Review by terrywbrady — COMMENTED** [review]
Without regex support, the DSpace 6 rest report tools will lose significant functionality.

Would it be possible to continue to provide this support through a deprecated method?  The REST Report tools are the only tools that use the OP enum operators.

File: dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java]: To clarify (as I misunderstood myself in Slack #dev discussion)... @terrywbrady is asking if we can:
1. Revert all changes to `OP` in this PR
2. Revert all changes to the `findByMetadataQuery()` method in this PR
3. Change the `findByMetadataQuery()` method to Deprecated (as it would still be dependent on deprecated Hibernate functionality). 

The reason for this is that `findByMetadataQuery()` is ONLY used by the existing, deprecated REST API (v6).  So, deprecating this method allows the deprecated REST API to work "as-is".  If we cannot find a replacement for this functionality in DSpace 7, then we simply drop the `findByMetadataQuery()` method -- or we implement a partial replacement (which may not include `matches` or `doesnt_match`)

Comment (terrywbrady) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java]: This looks good.

We should also 
- revert the change to rest.cfg

**Review by tdonohue — APPROVED** [review]

**Comment by tdonohue** [comment]
I've approved the PR as it has been fixed based on all changes I requests. However, I suggest we make the minor changes requested by @terrywbrady ☝️ prior to merging

**Comment by Raf-atmire** [comment]
@terrywbrady I've implemented a slightly altered version of the old findByMetadataQuery method, it includes the Hibernate logic again with the regex config. I've set the method to Deprecated and re-ran all the tests in DSpace 7 which passed. Could you test this method please? If it turns out that something is broken, I'll simply revert it back to the initial code to speed this PR up. Thanks!

**Review by terrywbrady — CHANGES_REQUESTED** [review]
@Raf-atmire , most of the functionality has been restored.  The exists and doesnt_exist operators are not currently working.  I think they are doing an equality test.  Exists should simply check if a metadata field is present (regardless of the value) for an item.  Doesnt_exist should verify that a metadata field does not exist for an item.

**Review by terrywbrady — APPROVED** [review]
I added a commit to address the outstanding issue I had reported.

**Comment by tdonohue** [comment]
Looks good, and this now has approval from several committers. Merging.  Thanks @Raf-atmire for your hard work on this!

**Pull request merged**
